File: quickstart/tests/test_config.py
"""Tests for quickstart.config — ProjectConfig dataclass."""

from __future__ import annotations

from pathlib import Path

from quickstart.config import ProjectConfig


class TestProjectConfig:
    """Tests for ProjectConfig.from_repo_root factory."""

    def test_from_repo_root_sets_device_dir(self, tmp_path: Path) -> None:
        config = ProjectConfig.from_repo_root(tmp_path)
        assert config.device_dir == tmp_path / "device"

    def test_from_repo_root_sets_cloud_dir(self, tmp_path: Path) -> None:
        config = ProjectConfig.from_repo_root(tmp_path)
        assert config.cloud_dir == tmp_path / "cloud"

    def test_from_repo_root_sets_env_dev(self, tmp_path: Path) -> None:
        config = ProjectConfig.from_repo_root(tmp_path)
        assert config.env_dev == tmp_path / ".env.dev"

    def test_from_repo_root_sets_env_staging(self, tmp_path: Path) -> None:
        config = ProjectConfig.from_repo_root(tmp_path)
        assert config.env_staging == tmp_path / ".env.staging"

    def test_default_sam_local_port(self, tmp_path: Path) -> None:
        config = ProjectConfig.from_repo_root(tmp_path)
        assert config.sam_local_port == 3000

    def test_default_sam_ready_timeout(self, tmp_path: Path) -> None:
        config = ProjectConfig.from_repo_root(tmp_path)
        assert config.sam_ready_timeout == 120

    def test_default_sam_config_envs(self, tmp_path: Path) -> None:
        config = ProjectConfig.from_repo_root(tmp_path)
        assert config.sam_config_env_dev == "dev"
        assert config.sam_config_env_prod == "prod"

    def test_frozen_dataclass_immutable(self, tmp_path: Path) -> None:
        config = ProjectConfig.from_repo_root(tmp_path)
        with __import__("pytest").raises(AttributeError):
            config.sam_local_port = 9999  # type: ignore[misc]

    def test_prod_api_endpoint(self, tmp_path: Path) -> None:
        config = ProjectConfig.from_repo_root(tmp_path)
        assert config.prod_api_endpoint == "https://supervictor.advin.io"

    def test_repo_root_stored(self, tmp_path: Path) -> None:
        config = ProjectConfig.from_repo_root(tmp_path)
        assert config.repo_root == tmp_path

File: supervictor/cli/tests/test_config.rs
use std::path::Path;

use qs::config::ProjectConfig;

#[test]
fn test_from_repo_root_paths() {
    let cfg = ProjectConfig::from_repo_root(Path::new("/tmp/repo"));
    assert_eq!(cfg.repo_root, Path::new("/tmp/repo"));
    assert_eq!(cfg.edge_dir, Path::new("/tmp/repo/supervictor/edge"));
    assert_eq!(cfg.env_dev, Path::new("/tmp/repo/.env.dev"));
    assert_eq!(cfg.env_staging, Path::new("/tmp/repo/.env.staging"));
    assert_eq!(cfg.env_prod, Path::new("/tmp/repo/.env.prod"));
    assert_eq!(cfg.log_dir, Path::new("/tmp/repo/.logs"));
}

#[test]
fn test_defaults() {
    let cfg = ProjectConfig::from_repo_root(Path::new("/tmp/repo"));
    assert_eq!(cfg.sam_local_port, 3000);
    assert_eq!(cfg.sam_ready_timeout, 120);
    assert_eq!(cfg.sam_config_env_dev, "dev");
    assert_eq!(cfg.sam_config_env_prod, "prod");
    assert_eq!(cfg.certs_dir_name, "certs");
    assert_eq!(cfg.prod_api_endpoint, "https://supervictor.advin.io");
}

#[test]
fn test_certs_dir() {
    let cfg = ProjectConfig::from_repo_root(Path::new("/tmp/repo"));
    assert_eq!(cfg.certs_dir(), Path::new("/tmp/repo/certs"));
}

#[test]
fn test_gen_certs_script_path() {
    let cfg = ProjectConfig::from_repo_root(Path::new("/tmp/repo"));
    assert_eq!(
        cfg.gen_certs_script_path(),
        Path::new("/tmp/repo/supervictor/endpoint/scripts/gen_certs.sh")
    );
}
